### PR TITLE
Add title to plot with fixed y coordinate of the root group.

### PR DESCRIPTION
**Pull request by bobocandys**

### CODE REVIEW

File: src/compile/compile.ts [hunk 1]
```diff
@@ -33,10 +33,10 @@ export function compile(spec) {
     config.viewport ? { viewport: config.viewport } : {},
     config.background ? { background: config.background } : {},
     keys(config.scene).length > 0 ? scene(config) : {},
-    {
-      data: compileData(model).concat([compileLayoutData(model)]),
-      marks: [compileRootGroup(model)]
-    });
+    { data: compileData(model).concat([compileLayoutData(model)]) },
```
Comment (domoritz) [inline src/compile/compile.ts]: Can you reformat this so it is easier to read?

File: src/compile/compile.ts [hunk 2]
```diff
@@ -57,6 +57,26 @@ function scene(config) {
   }, {});
 }
 
+export function compileTitle(model: Model) {
+  const spec = model.spec();
+  return {
```
Comment (domoritz) [inline src/compile/compile.ts]: no need to put `'` around keys

File: src/compile/compile.ts [hunk 3]
```diff
@@ -57,6 +57,26 @@ function scene(config) {
...
+      'name': 'title',
```
Comment (domoritz) [inline src/compile/compile.ts]: Please use the method in model to generate names

File: src/compile/compile.ts [hunk 4]
```diff
@@ -57,6 +57,26 @@ function scene(config) {
...
+      'type': 'text',
+      'from': {'data': 'layout'},
+      'properties': {
+        'update': {
+          'x': {'field': 'width', 'mult': 0.5},
+          'y': {'value': 0},
+          'text': {'value': spec.title},
+          'fill': {'value': 'black'},
```
Comment (domoritz) [inline src/compile/compile.ts]: All of these things should probably become configs. Can do later.

File: test/compile/compile.test.ts [hunk 5]
```diff
@@ -33,4 +33,38 @@ describe('Compile', function() {
     });
 
   });
+
+  describe('compileTitle()', function() {
+    describe('ordinal', function() {
+      const model = parseModel({
+        "description": "A simple bar chart with embedded data.",
```
Comment (domoritz) [inline test/compile/compile.test.ts]: Make the spec much smaller. Only add what you need.

File: test/compile/compile.test.ts [hunk 6]
```diff
@@ -33,4 +33,38 @@ describe('Compile', function() {
...
+      const titleGroup = compile.compileTitle(model);
+
+      it('should refer to layout data', function() {
+        assert.deepEqual(titleGroup.from, {"data": "layout"});
+      });
+
+      it('should have type of text', function() {
+        assert.deepEqual(titleGroup.type, "text");
+      });
+
```
Comment (domoritz) [inline test/compile/compile.test.ts]: Please tests that it actually sets the title ;-)

File: src/compile/compile.ts (hunk 3 above)
Comment (bobocandys) [inline src/compile/compile.ts]: Can you explain more on this? Use the method in model instead of giving `name: 'title` ?

File: src/compile/compile.ts (hunk 1 above)
Comment (kanitw) [inline src/compile/compile.ts]: One readable option is:

```
{ 
   data: compileData(model).concat([compileLayoutData(model)]),
   marks: [].concat(
     (spec.title ? [compileTitle(model)] : []),
     [compileRootGroup(model)]
   )
}
```

File: src/compile/compile.ts (hunk 3 above)
Comment (kanitw) [inline src/compile/compile.ts]:
```js
'name':   (model.name() ? model.name() + '-' : '') + 'title'
```

Comment (domoritz) [inline src/compile/compile.ts]: Hmm, I'm thinking whether we should change the `model.name` function to take a string parameter instead. I can do that outside this pr.

Comment (kanitw) [inline src/compile/compile.ts]: Yeah. Let's do that as another PR (as part of fixing #1103).

File: src/compile/compile.ts [hunk 7]
```diff
@@ -35,8 +35,12 @@ export function compile(spec) {
     keys(config.scene).length > 0 ? scene(config) : {},
     {
       data: compileData(model).concat([compileLayoutData(model)]),
-      marks: [compileRootGroup(model)]
-    });
+      marks: [].concat(
+       (spec.title ? [compileTitle(model)] : []),
```
Comment (kanitw) [inline src/compile/compile.ts]: For all `spec.title` reference, could you please add `model.title()` and call it?   (We plan to get rid of `model.spec()` so we don't want to call it in too many more places)

File: test/compile/compile.test.ts (hunk 6 above)
Comment (bobocandys) [inline test/compile/compile.test.ts]: I checked if it actually sets the correct title name. Is this okay?

File: src/compile/compile.ts (hunk 7 above)
Comment (bobocandys) [inline src/compile/compile.ts]: Done. It makes sense to do this.

File: src/compile/compile.ts [hunk 8]
```diff
@@ -68,10 +92,13 @@ export function compileRootGroup(model: Model) {
     {
       from: {data: LAYOUT},
       properties: {
-        update: {
-          width: {field: 'width'},
-          height: {field: 'height'}
-        }
+        update: extend(
+          model.title() !== undefined ? {y: {'value': 55} }: {},
```
Comment (domoritz) [inline src/compile/compile.ts]: 55 actually looked way to large for the example @kanitw showed today. Can you try different examples and see whether we need a more complex formula here?

**Comment by domoritz** [comment]
Closing since we will use composition.
